runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc

```diff
@@ -1238,23 +1403,88 @@ static iree_status_t iree_hal_vulkan_device_query_i64(
   iree_hal_vulkan_device_t* device = iree_hal_vulkan_device_cast(base_device);
   *out_value = 0;
 
-  if (iree_string_view_equal(category,
-                             iree_make_cstring_view("hal.executable.format"))) {
-    if (iree_string_view_equal(key,
-                               iree_make_cstring_view("vulkan-spirv-fb"))) {
+  if (iree_string_view_equal(category, IREE_SV("hal.executable.format"))) {
+    if (iree_string_view_equal(key, IREE_SV("vulkan-spirv-fb"))) {
       // Base SPIR-V always supported.
       *out_value = 1;
-    } else if (iree_string_view_equal(
-                   key, iree_make_cstring_view("vulkan-spirv-fb-ptr"))) {
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("vulkan-spirv-fb-ptr"))) {
       // SPIR-V with device addresses is optionally supported based on whether
       // we have device feature support.
       *out_value = iree_all_bits_set(
                        device->logical_device->enabled_features(),
                        IREE_HAL_VULKAN_FEATURE_ENABLE_BUFFER_DEVICE_ADDRESSES)
                        ? 1
                        : 0;
+      return iree_ok_status();
+    }
+  }
+
+  // Note that the device queries used here should match the ones used in
+  // mapToDeviceQuery() on the compiler side.
+  if (iree_string_view_equal(category, IREE_SV("hal.dispatch"))) {
+    if (iree_string_view_equal(key, IREE_SV("compute.f16"))) {
+      bool v = device->logical_device->supported_properties().compute_f16;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.f64"))) {
+      bool v = device->logical_device->supported_properties().compute_f64;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i8"))) {
+      bool v = device->logical_device->supported_properties().compute_i8;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i16"))) {
+      bool v = device->logical_device->supported_properties().compute_i16;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("compute.i64"))) {
+      bool v = device->logical_device->supported_properties().compute_i64;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("storage.8bit"))) {
+      bool v = device->logical_device->supported_properties().storage_8bit;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("storage.16bit"))) {
+      bool v = device->logical_device->supported_properties().storage_16bit;
+      *out_value = v ? 1 : 0;
+      return iree_ok_status();
+    }
+    if (iree_string_view_equal(key, IREE_SV("subgroup.arithmetic"))) {
```

Comment:
I commented on quinns PR (which I think overlaps a bit), but it'd be good to keep the number of unique queries as small as possible. Instead of the runtime needing to make dozens of queries to get 1 bit of information (executing thousands to tens of thousands of CPU instructions) you can combine these into logical fields - for example, one bitfield for supported compute types where bit 0 is i8, bit 1 is i16, bit 2 is i32, etc - then the compiler emits one query and can test the bits as it needs.

we don't want to go too far in that direction, but at least for when you find yourself with half a dozen of the same prefix of something it makes sense.

Reply:
SG. Merged queries of the same bucket into one.